dimod/higherorder.py

```diff
@@ -200,21 +283,68 @@ def _prod(iterable):
     return val
 
 
+def _prod_D(iterable, D):
+    val = [1.] * D
+    for v in iterable:
+        val *= v
+    return val
+
+
+def _shifted_poly(h, j, label_dict):
+    poly = {}
+    for k, v in h.items():
+        new_tup = (label_dict[k],)
+        poly[new_tup] = v
+    for k, v in j.items():
+        new_tup = tuple(sorted((label_dict[vidx] for vidx in list(k))))
+        poly[new_tup] = v
+    return poly
+
+
+def create_poly(h, J):
+    """ given h,J creates a single polynomial dict.
+    all h's will turn into tuples.
+
+    Args:
+        h (dict): a dict of linear variables
+        J (dict): a dict of quadratic and higher order variables
+
+    Returns
+        dict: a higher order problem dict containing both linear,
+    quadratic and n-order terms.
+
+    """
+
+    poly = {(k,): v for k, v in h.items()}
+    poly.update(J)
+    return poly
+
+
 def poly_energy(sample, poly):
     """Create a binary quadratic model from a higher order polynomial.
 
     Args:
-        sample (dict):
-            Sample for which to calculate the energy, formatted as a dict where keys
-            are variables and values are the value associated with each variable.
+        sample (dict or (list,np.array)):
+            Sample(s) for which to calculate the energy, formatted as a dict 
+            where keys are variables and values are the value associated with 
+            each variable. If formatted as a list or np.array, the terms in 
+            poly dict must be provided as variable order.
 
         poly (dict):
-            Polynomial as a dict of form {term: bias, ...}, where `term` is a tuple of
-            variables and `bias` the associated bias.
+            Polynomial as a dict of form {term: bias, ...}, where `term` is a 
+            tuple of variables and `bias` the associated bias.
 
     Returns:
-        float: The energy of the sample.
+        float or list: The energy of the sample(s).
 
     """
-    return sum(_prod(sample[v] for v in variables) * bias
-               for variables, bias in poly.items())
+
+    try:
+        _ = len(sample[0])
+        D = len(sample)
+        return sum(_prod_D([sample[:, v] for v in variables], D) * bias
+                   for variables, bias in poly.items())
+
+    except:
```

Comment:
This should catch a specific exception. Also it looks like you're using the `_ = len(sample[0])` as the test to decide which format to do. In that case you should just use an `if` branch.

Reply:
yeah this was me avoiding numpy...